v3/newrelic/instrumentation.go

```diff
@@ -36,16 +36,32 @@ func WrapHandle(app *Application, pattern string, handler http.Handler, options
 	if app == nil {
 		return pattern, handler
 	}
+
 	// add the wrapped function to the trace options as the source code reference point
-	// (to the beginning of the option list, so that the user can override this)
-	// If we came here via WrapHandleFunc, that means it will have pushed the handler's
-	// code location if possible, so the options will be (this code location), (handler function location),
-	// (other options); that results in having a backup in case either failed, with preference
-	// given to the handler's function location. In any event, the user may override all of that
-	// with a location they choose.
+	// (but only if we know we're collecting CLM for this transaction and the user didn't already
+	// specify a different code location explicitly).
+
 	return pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		options = append([]TraceOption{WithThisCodeLocation()}, options...)
-		txn := app.StartTransaction(r.Method+" "+pattern, options...)
+		var tOptions *traceOptSet
+		var txnOptionList []TraceOption
+
+		if app.app != nil && app.app.run != nil && app.app.run.Config.CodeLevelMetrics.Enabled {
+			tOptions = resolveCLMTraceOptions(options)
+			if tOptions != nil && !tOptions.SuppressCLM && (tOptions.DemandCLM || app.app.run.Config.CodeLevelMetrics.Scope == 0 || (app.app.run.Config.CodeLevelMetrics.Scope&TransactionCLM) != 0) {
+				// we are for sure collecting CLM here, so go to the trouble of collecting this code location.
+				if tOptions.LocationOverride == nil {
+					WithThisCodeLocation()(tOptions)
+				}
+			}
+		}
+		if tOptions == nil {
+			// we weren't able to curate the options above, so pass whatever we were given downstream
+			txnOptionList = options
+		} else {
+			txnOptionList = append(txnOptionList, withPreparedOptions(tOptions))
```

Comment:
I think this can be simplified:
```go
txOptionList = []TraceOption{ func(o *traceOptSet) { *o = *tOptions } }
```

Reply:
At the moment that's equivalent and a tiny bit faster to execute, and if I don't end up needing the more nuanced implementation of `withPreparedOptions` which won't do things like overwrite a non-empty slice of values with an empty one from the incoming struct, I'll probably refactor that back to what's suggested here.
And it may likely turn out that the `withPreparedOptions` thing isn't needed. I'm experimenting with a few ideas here and refactoring out what doesn't seem to add value.